### PR TITLE
Improve the help message for an invalid calling convention

### LINKED ISSUE #93601: The help message when an invalid calling convention is specified is getting a little unwieldy
After #93561 misspelling a calling convention will output the following note:

```
LL | extern "路濫狼á́́" fn foo() {}
   |        ^^^^^^^^^ invalid ABI
   |
   = help: valid ABIs: Rust, C, C-unwind, cdecl, cdecl-unwind, stdcall, stdcall-unwind, fastcall, fastcall-unwind, vectorcall, vectorcall-unwind, thiscall, thiscall-unwind, aapcs, aapcs-unwind, win64, win64-unwind, sysv64, sysv64-unwind, ptx-kernel, msp430-interrupt, x86-interrupt, amdgpu-kernel, efiapi, avr-interrupt, avr-non-blocking-interrupt, C-cmse-nonsecure-call, wasm, system, system-unwind, rust-intrinsic, rust-call, platform-intrinsic, unadjusted
```

This is getting pretty hard to read!

There are a couple of improvements we could make:

1. Invoke Levenshtein magic and only suggest some calling conventions that are close, if any (covers mis-spellings);
    * This could be a `MachineApplicable::MaybeIncorrect` rustfix!
2. Suggest at most a couple of the most common ABIs (possibly a target-specific list?);
3. Do not suggest CCs that cannot be used due to them being unstable and feature not being enabled; and finally
4. Explain where to find a full list (may involve adding CLI surface to rustc).

With these suggestions in place the message could look like:

```
LL | extern "wni64" fn foo() {}
   |        ^^^^^^^ invalid ABI
   |
   = help: did you mean `win64`?
   = note: invoke `rustc --print=calling-conventions` for a full list of supported calling conventions
```

or

```
LL | extern "路濫狼á́́" fn foo() {}
   |        ^^^^^^^^^ invalid ABI
   |
   = help: commonly used calling conventions include: `C`, `system`, `sysv64`, `win64` and `wasm`
   = note: invoke `rustc --print=calling-conventions` for a full list of supported calling conventions
```

### PR DESCRIPTION
**Pull request by khyperia**
Fixes https://github.com/rust-lang/rust/issues/93601

I mostly followed the suggestions of @nagisa in that issue, ~~however, I wasn't sure how to check stability for the suggestion of "Do not suggest CCs that cannot be used due to them being unstable and feature not being enabled", so I did not implement that point.~~

I haven't contributed to rustc much, please feel free to point out suggestions! For example, the `.map(|s| Symbol::intern(s)).collect::<Vec<_>>()` seems pretty gross performance-wise, but maybe that's OK in error reporting code.

### CODE REVIEW

**Review by jyn514 — COMMENTED** [review]
> I wasn't sure how to check stability

I'm not sure either, but I'd suggest compiling a crate with an unstable ABI and using -Ztreat-err-as-bug to see where the error is reported. Then you can use that same logic in the diagnostics code :)

> the .map(|s| Symbol::intern(s)).collect::<Vec<_>>() seems pretty gross performance-wise, but maybe that's OK in error reporting code.

Yeah, that's fine, error reporting doesn't care about performance.

File: compiler/rustc_driver/src/lib.rs
Comment (khyperia) [reply compiler/rustc_driver/src/lib.rs:723]: Hmm, I'm not finding any tests for other --print options, e.g. --print=target-list has extremely similar code to this one. Could you point out where those are so I can add this test alongside them?

**Comment by khyperia** [comment]
> I'm not sure either, but I'd suggest compiling a crate with an unstable ABI and using -Ztreat-err-as-bug to see where the error is reported. Then you can use that same logic in the diagnostics code :)

Hmm. The errors seem to be reported here https://github.com/rust-lang/rust/blob/75b7e52e92c3b00fc891b47f5b2efdff0a2be55a/compiler/rustc_ast_passes/src/feature_gate.rs#L77

The crate rustc_ast_passes is not available in rustc_ast_lowering. Should I add a reference? Also, I'm not sure how to call that function without reporting an error - do I need to construct a dummy Session or something like that? How do I do that?

**Comment by jyn514** [comment]
> The crate rustc_ast_passes is not available in rustc_ast_lowering. Should I add a reference? Also, I'm not sure how to call that function without reporting an error - do I need to construct a dummy Session or something like that?

I'd suggest taking that giant match and refactoring it a bit - add a public function in rustc_ast_lowering that only determines whether the ABI is unstable or not, then calling it from the feature gate code in rustc_ast_passes. Then you don't need a dummy session and you can make sure the diagnostic code uses the same logic as the feature gate code.

File: compiler/rustc_driver/src/lib.rs
Comment (jyn514) [reply compiler/rustc_driver/src/lib.rs:723]: Hmm, not sure off the top of my head. Maybe it's in `run-make`? If I were at a laptop I'd find this with `grep -r -- --print src/test`, but I'm on mobile right now.

Comment (khyperia) [reply compiler/rustc_driver/src/lib.rs:723]: Ah thanks! Found this, but it's a bit odd as there seems to be no expected output. I'll add a similar Makefile for `calling-conventions`. https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/print-target-list

File: compiler/rustc_target/src/spec/abi.rs
Comment (jyn514) [inline compiler/rustc_target/src/spec/abi.rs:112]: nit: I would change this from a macro to a function, since it no longer affects control flow.

**Review by jyn514 — APPROVED** [review]
This looks great :) I'll let a compiler member review since I don't often touch target specs, but the code LGTM.

File: compiler/rustc_target/src/spec/abi.rs
Comment (khyperia) [reply compiler/rustc_target/src/spec/abi.rs:112]: The one slight annoyance with that is that `$feature` is used in both `$features.$feature` (bool field for if it's enabled), as well as `sym::$feature`, so the caller would have to duplicate the name of the feature, passing in both. I guess that's okay, though, and probably worth it being a regular function over a macro?

Comment (jyn514) [reply compiler/rustc_target/src/spec/abi.rs:112]: Hmm. I agree changing it is a pain. But I also kind of feel that if this is a macro, it's not helping *enough* for it to be worth it - it would be nice to find a way to simplify the giant match some more, e.g. only pass in a list of ABI names and feature gates and have it generate the match for you.

That said, it's a big change unrelated to your improvement, I'm fine with just merging as is.

Comment (khyperia) [reply compiler/rustc_target/src/spec/abi.rs]: ah! @m-ou-se pointed out that `Features::enabled(Symbol)` exists, that makes it nicer.

File: src/test/run-make-fulldeps/print-calling-conventions/Makefile
Comment (nagisa) [inline src/test/run-make-fulldeps/print-calling-conventions/Makefile:4]: Hm, wouldn’t this be better off as a UI test?

**Review by nagisa — COMMENTED** [review]
This looks great!

**Comment by nagisa** [comment]
> I'll let a compiler member review since I don't often touch target specs, but the code LGTM.

r=me in that regard. This needs a rebase and this could probably benefit from fewer commits (e.g. squash in the commits that fix mistakes introduced by prior commits as well as those that add tests)

File: src/test/ui/parser/issues/issue-8537.stderr
Comment (khyperia) [reply src/test/ui/parser/issues/issue-8537.stderr:7]: I'm not sure! If you have a better suggestion, feel free, I'm just following your exact wording in #93601

File: src/test/run-make-fulldeps/print-calling-conventions/Makefile
Comment (khyperia) [reply src/test/run-make-fulldeps/print-calling-conventions/Makefile:4]: Not sure - I'm following the example set by other similar tests. If we want to change the group of them to test in a different way, that seems like a separate change IMO https://github.com/rust-lang/rust/pull/100488#discussion_r945289589

**Comment by khyperia** [comment]
I'm not sure how to resolve the merge conflict from https://github.com/rust-lang/rust/commit/0043d10c712769b45d6cb7fb3fcc141878263d18, to be honest. I don't know how this new error reporting system works, or what this .ftl file is. Is there any documentation on how this extremely magical-seeming system works, so I can use it properly? (macros are one thing, this seems like a whole custom DSL, haha)

**Comment by jyn514** [comment]
cc @davidtwco , is there documentation somewhere for how to use the translation macros?

**Comment by davidtwco** [comment]
> cc @davidtwco , is there documentation somewhere for how to use the translation macros?

https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-structs.html primarily, #100717 links most of our resources

**Comment by khyperia** [comment]
Sorry for taking so long for the update! @nagisa could I get another review? I (...we) changed some stuff in the rebase, it might be nice to check again. Thanks!

**Comment by jyn514** [comment]
@khyperia fyi you can use

@rustbot ready

to mark this as ready for review.

**Comment by nagisa** [comment]
@bors r-
@bors r=nagisa,jyn514

(/me wonders if bors can stomach this...)

**Pull request merged**
